platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemSpec.java

```diff
@@ -122,6 +123,42 @@ public interface ProblemSpec {
      */
     ProblemSpec solution(String solution);
 
+    /**
+     * Declares additional data attached to the problem.
+     *
+     * @param type The type of the additional data.
+     * This can be any type that implements {@link AdditionalData} including {@code abstract} classes and interfaces.
+     * This type will be instantiated and provided as an argument for the {@code Action} passed as the second argument.
+     *
+     * <p>The limitations for this type are:</p>
+     * <ul>
+     *  <li>Only {@code get<VALUE>} and {@code set<VALUE>} methods are allowed.
+     *  <li>These are only allowed to use these types:
+     *    <ul>
+     *         <li>{@code String}</li>
+     *         <li>{@code Boolean}</li>
+     *         <li>{@code Character}</li>
+     *         <li>{@code Byte}</li>
+     *         <li>{@code Short}</li>
+     *         <li>{@code Integer}</li>
+     *         <li>{@code Float}</li>
+     *         <li>{@code Long}</li>
+     *         <li>{@code Double}</li>
+     *         <li>{@code BigInteger}</li>
+     *         <li>{@code BigDecimal}</li>
+     *         <li>{@code File}</li>
+     *   </ul>
+     * </ul>
```

Comment:
We do not enforce that yet, but please document, that clients should also expect a runtime exception when different additional data types are reported for the same `problemId`.

Reply:
I'll do this

`* @throws IllegalArgumentException if the conditions for the type are not met or if a different type for the same problem id is used.`